cli/src/commands/run.rs

```
@@ -14,12 +14,262 @@
+    NoDirectoryCreated,
+    #[error("the command failed {} for {}")]
+    CommandFailure(ExitCode, CommitId),
+}
+
+impl From<RunError> for CommandError {
+    fn from(value: RunError) -> Self {
+        CommandError::new(crate::command_error::CommandErrorKind::Cli, Box::new(value))
+    }
+}
+
+/// Creates the required directories for a StoredWorkingCopy.
+/// Returns a tuple of (`output_dir`, `working_copy` and `state`).
+fn create_working_copy_paths(
+    path: &PathBuf,
+) -> Result<(PathBuf, PathBuf, PathBuf), std::io::Error> {
+    let output = path.join("output");
+    let working_copy = path.join("working_copy");
+    let state = path.join("state");
+    std::fs::create_dir(&output)?;
+    std::fs::create_dir(&working_copy)?;
+    std::fs::create_dir(&state)?;
+    Ok((output, working_copy, state))
+}
+
+/// Represent a `MergeTreeId` in a way that it may be used as a working-copy
+/// name. This makes no stability guarantee, as the format may change at
+/// any time.
+fn to_wc_name(id: &MergedTreeId) -> String {
+    match id {
+        MergedTreeId::Legacy(tree_id) => tree_id.hex(),
+        MergedTreeId::Merge(tree_ids) => {
+            let ids = tree_ids
+                .map(|id| id.hex())
+                .iter_mut()
+                .enumerate()
+                .map(|(i, s)| {
+                    // Incredibly "smart" way to say, append "-" if the number is odd "+"
+                    // otherwise.
+                    if i & 1 != 0 {
+                        s.push('-');
+                    } else {
+                        s.push('+');
+                    }
+                    s.to_owned()
+                })
+                .collect_vec();
+            let mut obfuscated: String = ids.concat();
+            // `PATH_MAX` could be a problem for different operating systems, so truncate
+            // it.
+            if obfuscated.len() >= 255 {
+                obfuscated.truncate(200);
+            }
+            obfuscated
```

issue: I would rather just have the filesystem operation fail, rather than silently truncate the full path. In particular, it would probably be possible to add more sides to a conflict and then have two unrelated trees get assigned the same path on disk.

---

This works for *Nix based systems but for Windows where [this option](https://learn.microsoft.com/en-us/windows/win32/fileio/maximum-file-path-limitation?tabs=registry#enable-long-paths-in-windows-10-version-1607-and-later) is not set, so it'll make the command unusable. So maybe there's a better solution?

---

I commented out the code for now. If that is something we want, I'll remove it.

---

its deleted now.